docs/spec/spec.md

```diff
@@ -1459,10 +1468,166 @@ parseHeader(string headerValue) returns HeaderValue[]|ClientError  {
+service class RequestInterceptor {
+   *http:RequestInterceptor;
+ 
+   resource function 'default [string… path](http:FilterContext ctx, http:Caller caller,
+                       http:Request req) returns error? {
+       request.setHeader("X-requestHeader", "RequestInterceptor");
+       ctx.next();
+   }
+}
+```
+
+Since interceptors work with network activities, it must be either a remote or resource function. In this case resource 
+functions are used for RequestInterceptor as it gives more flexibility. With resource functions interceptors can be engaged 
+based on HTTP method and path.
+
+For instance consider a scenario where there are two resources; one on path foo whereas the other on path bar. If the 
+user writes a interceptor as follows, it would only get hit when the request is directed to foo resource.
+
+```ballerina
+service class RequestInterceptor {
+   *http:RequestInterceptor;
+ 
+   resource function 'default foo(http:FilterContext ctx, http:Caller caller,
+                       http:Request req) returns error? {
+       request.setHeader("X-requestHeader", "RequestInterceptor");
+       ctx.next();
+   }
+}
+```
+
+##### 8.1.1.1 FilterContext  
+Following is the rough definition of the interceptor context.
+```ballerina
+public isolated class FilterContext {
+    private final map<value:Cloneable|isolated object {}> attributes = {};
+
+    public isolated function add(string 'key, value:Cloneable|isolated object {} value) {
+        if value is value:Cloneable {
+            lock {
+                self.attributes['key] = value.clone();
+            }
+        }
+        else {
+            lock {
+                self.attributes['key] = value;
+            }   
+        }
+    }
+
+    public isolated function get(string 'key) returns value:Cloneable|isolated object {} {
+        lock {
+            return self.attributes.get('key);
+        }
+    }
+
+    public isolated function remove(string 'key) {
+        lock {
+            value:Cloneable|isolated object {} _ = self.attributes.remove('key);
+        }
+    }
+
+    public isolated function next() returns error? = external;
+}
+```
+
+##### 8.1.1.2 next() Method  
+However, there is an addition when it comes to FilterContext. A new method namely, next() is introduced to control 
+the execution flow. Users must invoke next() method in order to trigger the next interceptor in the pipeline. 
+Previously, this was controlled by returning a boolean value which is quite cryptic and confusing.
+
+Moreover, invoking next() method results in immediately executing the next interceptor in the pipeline. Which means 
+whatever below the next() method will only get executed after the completion of execution of the next Interceptor.
+
+If the users forget to call next() and don't respond either, the request will be left hanging.
+
+##### 8.1.1.3 Returning error?
+Resource functions can return only error?. In the case of an error, interceptor chain execution jumps to the last 
```

Comment:
Resource functions can return `NextService|error?`. They can return only `error?` **when using the `Caller`**. 
Also, in case of an error, interceptor pipeline execution jumps to the **nearest `RequestErrorInterceptor`** in the pipeline. This `RequestErrorInterceptor` is not necessarily the last interceptor in the pipeline, it can be anywhere in the chain.